### PR TITLE
Batch log enqueue and dequeue operations

### PR DESCRIPTION
**Pull request by chaudum**
**What this PR does / why we need it**:

To have better observability of the scheduler and the hierarchical queues, this PR adds logging for otherwise high cardinality metrics.

* Enqueue operations per tenant and actor path
* Dequeue operations per tenant and querier

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Review by owen-d — COMMENTED** [review]
I'm not in love with adding a mutex lock into scheduler enqueuing for this, but I understand it helps to deamplify log lines. Maybe we should hold off on this and see if we actually need it.

**Comment by chaudum** [comment]
> I'm not in love with adding a mutex lock into scheduler enqueuing for this, but I understand it helps to deamplify log lines. Maybe we should hold off on this and see if we actually need it.

Yeah, understand the concern. I moved the PR back to draft.

File: pkg/scheduler/queue/queue.go [hunk 1]
```diff
@@ -128,7 +128,7 @@ func (q *RequestQueue) Enqueue(tenant string, path []string, req Request, maxQue
 // Dequeue find next tenant queue and takes the next request off of it. Will block if there are no requests.
 // By passing tenant index from previous call of this method, querier guarantees that it iterates over all tenants fairly.
 // If querier finds that request from the tenant is already expired, it can get a request for the same tenant by using UserIndex.ReuseLastUser.
-func (q *RequestQueue) Dequeue(ctx context.Context, last QueueIndex, querierID string) (Request, QueueIndex, error) {
+func (q *RequestQueue) Dequeue(ctx context.Context, last QueueIndex, querierID string) (Request, QueueIndex, string, error) {
```
Comment (jeschkies) [inline pkg/scheduler/queue/queue.go:131]: Could you add a doc string on what's returned?

File: pkg/scheduler/util/batchlogger.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/services"
+)
+
+type BatchLogger struct {
```
Comment (jeschkies) [inline pkg/scheduler/util/batchlogger.go:12]: Could you add a doc string explaining what it does?

**Pull request closed without merging**